---
title: "SimusCopR"
output: html_document
---

Wrap-up R package to coordinate SimusCop simulations. 

Also, here we present a easy conversion of VCF files in variations and snp files for SimuSCoP simuReads. This tools is useful if you want to simulate Illumina reads based on empirical variations already stored in a VCF file.

Just use Docker!

# Input files

All inputs must have information for just one chromosome and one sample.

* Reference genome and indexes from picard, bwa and samtools
* Sorted BAM file from real data
* VCF file
* BED file

If you need to filter the files to keep only one chromosome, we suggest:

```{bash, eval=FALSE}
docker run -v $(pwd):/ cristaniguti/r-samtools samtools view -b /your_bam_file Chr10 > filtered_bam

docker run -v $(pwd):/ cristaniguti/vcftools vcftools --gzvcf /vcf.file.gz --indv ID1 --chr Chr01 --recode --out vcf_filt
```

To obtain the BED file from the BAM file

```{bash}
docker run -it -v $(pwd):/data  biocontainers/bedtools:v2.28.0_cv1 bamtobed -i filtered_bam > bed_file
```

# Create profile

```{r}
seqToProfile("PT_F.sorted.bam", "PT_F.sorted_filt.bed", "gatk_chr10_PT_F.recode.vcf",
             "references/Chr10.populus.fa", "PT_F.profile")

AdaptBed(bed.file = "PT_F.sorted_filt.bed", out.file = "PT_F.sorted_filt_edit2.bed", int.size = 500)
```

# Variants files

```{r}
library(vcfR)

vcfR.object <- read.vcfR("gatk_chr10_PT_F.recode.vcf")

variants <- vcf2variants(vcfR.object, sample = "PT_F", chrom = "Chr10")

write.table(variants$SNVs, file = "SNVs.txt", sep = "\t", quote = F, col.names = F, row.names = F)
write.table(variants$indels, file = "indels.txt", sep = "\t", quote = F, col.names = F, row.names = F)
write.table(variants$insertions, file = "insertions.txt", sep = "\t", quote = F, col.names = F, row.names = F)

system("cat SNVs.txt indels.txt insertions.txt > variants.txt")
```

# Run simulation

```{r}
simuReads(ref = "references/Ptrichocarpa_533_v4.0.fa",
          profile = "PT_F.profile",
          variation = "variants.txt",
          target = "PT_F.sorted_filt_edit2.bed",
          name = "PT_F",
          output = "./results",
          layout = "SE",
          threads = 6,
          verbose = 1,
          coverage = 15)
```


# Test to see if worked

```{bash}
docker run -v $(pwd):/data  kfdrc/bwa-picard:latest-dev  bin/bwa mem -t 2  \
data/references/Ptrichocarpa_533_v4.0.fa data/results/P1_100_150_1_2enz.fq.gz | \
java -jar /picard.jar SortSam \
I=/dev/stdin \
O="data/results/P1_2enz.bam" \
TMP_DIR=./tmp \
SORT_ORDER=coordinate \
CREATE_INDEX=true

# Images on IGV
```


# Simulation with Reseq


