---
title: "SimusCopR"
output: html_document
---

Wrap-up R package to coordinate SimusCop simulations. 

Also, here we present a easy conversion of VCF files in variations and snp files for SimuSCoP simuReads. This tools is useful if you want to simulate Illumina reads based on empirical variations already stored in a VCF file.

# Create profile

```{bash}
# Parent 1

# Inputs
SAMPLE=PT_F
BAM=${SAMPLE}.sorted.bam
FASTA=references/Ptrichocarpa_533_v4.0.fa
FASTA_CHR=references/Chr10.populus.fa #and indexes
VCF=gatk.vcf.gz

# Outputs
BAM_CHR=${SAMPLE}.chr.sorted.bam
VCF_FILT=gatk_chr10_${SAMPLE}.recode.vcf
BED=${SAMPLE}.sorted.bed
BED_FILT=${SAMPLE}.sorted_filt.bed
PROFILE=${SAMPLE}.profile
BED_FILT_EDIT=${SAMPLE}.sorted_filt_edit.bed

samtools view -b ${BAM} Chr10 > ${BAM_CHR}

#vcftools --gzvcf gatk.vcf.gz --indv PT_F --chr Chr10 --recode --out gatk_chr10_PT_F

docker run -it -v $(pwd):/data  biocontainers/bedtools:v2.28.0_cv1 bamtobed -i ${BAM_CHR} > ${BED_FILT}

#docker run -it -v $(pwd):/data biocontainers/bedops:v2.4.35dfsg-1-deb_cv1 bedextract Chr10 ${BED} > ${BED_FILT}

bin/seqToProfile -b ${BAM_CHR} -t ${BED_FILT} -v ${VCF_FILT} -r ${FASTA_CHR} > ${PROFILE}
```

```{r}
AdaptBed(bed.file = "PT_F.sorted_filt.bed", out.file = "PT_F.sorted_filt_edit2.bed", int.size = 500)
```

# Input files

Didn't find CNV in HaplotypeCaller result. The script will convert SNVs, indels and deletions.

Create pedigreesim inputs

```{r}
create_founderfiles <- function(vcfR.object, parents_id, chrom){
  if(!is(vcfR.object, "vcfR")) stop("Input object is not of class vcfR")
  
  ALT <- vcfR.object@fix[,5]
  
  # Remove markers with more than one alternative
  rm_mks <- grep(ALT, pattern = ",")
  ALT <- ALT[-rm_mks]
  REF <- vcfR.object@fix[,4][-rm_mks]
  CHROM <- vcfR.object@fix[,1][-rm_mks]
  POS <- vcfR.object@fix[,2][-rm_mks]
  GT <- vcfR.object@gt[-rm_mks,]
  
  if(!is.null(chrom)){
    chr <- which(CHROM %in% chrom)
    ALT <- ALT[chr]
    REF <- REF[chr]
    CHROM <- CHROM[chr]
    POS <- POS[chr]
    GT <- GT[chr,]
  }
  
  # Genotypes
  GT <- GT %>% as.data.frame %>% select(all_of(parents_id))  %>% 
    apply(.,2,as.character) %>% strsplit(., ":") %>% sapply(., "[[",1) %>% 
    matrix(., ncol=length(samples)) %>% 
    gsub(pattern = "[|]", replacement = "/") 
  
  colnames(GT) <- parents_id
  
  P1 <- GT %>% as.data.frame %>% select(.,parents_id[1]) 
  P1 <- strsplit(as.character(P1[,1]), "/")
  P1_1 <- sapply(P1, "[[", 1)
  P1_2 <- sapply(P1, "[[", 2)
  P1_1[P1_1 == "0"] <- REF[P1_1 == "0"]
  P1_1[P1_1 == "1"] <- ALT[P1_1 == "1"]
  P1_2[P1_2 == "0"] <- REF[P1_2 == "0"]
  P1_2[P1_2 == "1"] <- ALT[P1_2 == "1"]
  
  
  P2 <- GT %>% as.data.frame %>% select(.,parents_id[2]) 
  P2 <- strsplit(as.character(P2[,1]), "/")
  P2_1 <- sapply(P2, "[[", 1)
  P2_2 <- sapply(P2, "[[", 2)
  P2_1[P2_1 == "0"] <- REF[P2_1 == "0"]
  P2_1[P2_1 == "1"] <- ALT[P2_1 == "1"]
  P2_2[P2_2 == "0"] <- REF[P2_2 == "0"]
  P2_2[P2_2 == "1"] <- ALT[P2_2 == "1"]
  
  # Phases in parents will be the same as the VCF
  founderfile <- tibble(
    marker = str_c(CHROM, POS, sep = "_"),
    P1_1,
    P1_2,
    P2_1,
    P2_2
  ) 
  
  return(founderfile)
}
```



```{r}
library(vcfR)

vcfR.object <- read.vcfR("gatk_chr10_PT_F.recode.vcf")

variants <- vcf2variants(vcfR.object, sample = "PT_F", chrom = "Chr10")

write.table(variants$SNVs, file = "SNVs.txt", sep = "\t", quote = F, col.names = F, row.names = F)
write.table(variants$indels, file = "indels.txt", sep = "\t", quote = F, col.names = F, row.names = F)
write.table(variants$insertions, file = "insertions.txt", sep = "\t", quote = F, col.names = F, row.names = F)

system("cat SNVs.txt indels.txt insertions.txt > variants.txt")
```

# Run simulation

# File config_test.txt

```{bash}
##------arguments for generating sequencing reads------##

## [required] fasta reference file from which reads will be sampled
ref = references/Ptrichocarpa_533_v4.0.fa

## [required] a model file generated by "seqToProfile" utility
profile = PT_F.profile

## [optional] files defining the variations (indel, SNV and CNV) to simulate
variation = variants.txt

## [optional] file defining the SNPs to simulate
# snp = ./testData/snp.txt

## [optional] file defining target regions for sequencing. If the argument is not specified, sequencing data of all chromosomes defined in the reference will be produced. 
target = PT_F.sorted_filt_edit2.bed

## [required] popolation names (comma-separated).
name = PT_F

## [required] output directory
output = ./results

## [optional] sequence layout, should be "SE" for single end or "PE" for paired-end (default: "SE")
layout = SE

## [optional] the number of threads to use (default: 1)
threads = 6

## [optional] print the intermediate information, should be 0 or 1 (default: 1)
verbose = 1

## [required] sequencing coverage
coverage = 15

## [optional] insert size for paired end sequencing (default: 350, only effective when the "layout" is set to "PE")
## insertSize = 200
```

```{bash}
bin/simuReads config_test.txt
```

# Test to see if worked

```{bash}
docker run -v $(pwd):/data  kfdrc/bwa-picard:latest-dev  bin/bwa mem -t 2  \
data/references/Ptrichocarpa_533_v4.0.fa data/results/P1_100_150_1_2enz.fq.gz | \
java -jar /picard.jar SortSam \
I=/dev/stdin \
O="data/results/P1_2enz.bam" \
TMP_DIR=./tmp \
SORT_ORDER=coordinate \
CREATE_INDEX=true

# Images on IGV
```


# Simulation with Reseq


