#' Create SimusCop profile
#' 
#' @param ref fasta reference file from which reads will be sampled
#' @param profile a model file generated by "seqToProfile" utility
#' @param variation files defining the variations (indel, SNV and CNV) to simulate
#' @param snp file defining the SNPs to simulate
#' @param target file defining target regions for sequencing. If the argument is not specified, sequencing data of all chromosomes defined in the reference will be produced. 
#' @param name popolation names (comma-separated).
#' @param output output directory
#' @param layout sequence layout, should be "SE" for single end or "PE" for paired-end (default: "SE")
#' @param threads the number of threads to use (default: 1)
#' @param verbose print the intermediate information, should be 0 or 1 (default: 1)
#' @param coverage sequencing coverage
#' @param insertSize insert size for paired end sequencing (default: 350, only effective when the "layout" is set to "PE")
#' 
#' @export
simuReads <- function(ref = NULL, 
                      profile = NULL, 
                      variation = NULL, 
                      target = NULL, 
                      name = NULL, 
                      output = NULL, 
                      layout = NULL,
                      threads = NULL, 
                      verbose = NULL, 
                      coverage = NULL, 
                      insertSize = NULL){
  
  configs <- c(ref=ref, profile=profile,variation=variation, target=target, 
               name=name, output=output, layout=layout, 
               threads=threads, verbose=verbose, coverage=coverage, insertSize=insertSize)
  
  temp <- tempfile()
  write.table(configs, quote = F, row.names = T, file = temp, col.names = F, sep = "=")
  
  system(
    paste0(system.file(package="simuscopR"),"/SimusCop/simuReads ", temp)
  )
}
